lib.php

```diff
@@ -216,67 +229,91 @@ function block_coursefeedback_insert_question($question, $feedbackid, $questioni
-                "newpos" => $newpos,
-            )
-        );
-        // Move to new position.
-        $sql[] = array(
-            "query" => "UPDATE {block_coursefeedback_questns}
-                           SET questionid = :newpos, timemodified = :modified
-                         WHERE coursefeedbackid = :fid 
-                               AND questionid = :oldpos",
-            "params" => array(
-                "newpos" => $newpos,
-                "modified" => time(),
-                "fid" => $feedbackid,
-                "oldpos" => $oldpos,
-            )
-        );
-        // Restore old position.
-        $sql[] = array(
-            "query" => "UPDATE {block_coursefeedback_questns}
-                           SET questionid = :oldpos, timemodified = :modified
-                         WHERE coursefeedbackid = :fid 
-                               AND questionid = :tmppos",
-            "params" => array(
-                "oldpos" => $oldpos,
-                "modified" => time(),
-                "fid" => $feedbackid,
-                "tmppos" => $tmppos,
-            )
-        );
+    if ($oldposition === $newposition) {
+        return true;
+    }
 
-        return block_coursefeedback_execute_sql_arr($sql);
-    } else {
-        return false;
+    $transaction = $DB->start_delegated_transaction();
+
+    try {
+        // Get the current offset value.
+        $offset = block_coursefeedback_get_questionid($feedbackid);
+
+        // Step 1: Shift all affected questions to a temporary offset.
+        $sql = "UPDATE {block_coursefeedback_questns}
+                   SET questionid = questionid + :offset
+                 WHERE coursefeedbackid = :feedbackid
+                       AND questionid BETWEEN LEAST(:oldposition::bigint, :newposition::bigint)
+                                     AND GREATEST(:oldposition1::bigint, :newposition1::bigint)";
+        $DB->execute($sql, [
+            'offset'      => $offset,
+            'feedbackid'  => $feedbackid,
+            'oldposition' => $oldposition,
+            'newposition' => $newposition,
+            'oldposition1' => $oldposition,
+            'newposition1' => $newposition,
+        ]);
+
+        // Step 2: Adjust positions of intermediate questions.
+        if ($oldposition < $newposition) {
+            // Moving to a higher position: shift intermediate questions one step down, considering the offset.
+            $sql = "UPDATE {block_coursefeedback_questns}
+                       SET questionid = questionid - (1 + :offset)
+                     WHERE coursefeedbackid = :feedbackid
+                           AND questionid BETWEEN (:oldposoffset + 1) AND :newposoffset";
+            $DB->execute($sql, [
+                'newposoffset' => $newposition + $offset,
+                'oldposoffset' => $oldposition + $offset,
+                'offset'       => $offset,
+                'feedbackid'   => $feedbackid,
+            ]);
```

Comment:
Since you repeat the same call with the same array argument below, you could move the `$DB->execute(...)` call out of and under the condition blocks.